### PR TITLE
Cleanup and rewrite some symmetric functions

### PR DESCRIPTION
**Pull request by fredrikekre**
https://gist.github.com/fredrikekre/b14fc545a10dc532835a92ea044f3187

### CODE REVIEW

File: src/ContMechTensors.jl [hunk 1]
```diff
@@ -42,9 +42,6 @@ typealias SymmetricTensors{dim, T} Union{SymmetricTensor{2, dim, T}, SymmetricTe
 typealias Tensors{dim, T} Union{Tensor{2, dim, T}, Tensor{4, dim, T}, Vec{dim, T}}
 
 include("utilities.jl")
-include("tuple_utils.jl")
```
Comment (fredrikekre) [inline src/ContMechTensors.jl:45]: gathered all remaining utils in `utilities.jl`

File: src/tensor_products.jl [hunk 2]
```diff
@@ -42,43 +42,44 @@ end
...
+    for i in 1:dim, j in i:dim
+        if i == j
+            push!(ex, :(get_data(S1)[$(idx2(i, j))] * get_data(S2)[$(idx2(i, j))]))
         else
-            @code :(@inbounds s += 2 * data1[$k] * data2[$k])
+            push!(ex, :(2 * get_data(S1)[$(idx2(i, j))] * get_data(S2)[$(idx2(i, j))]))
         end
     end
-    @code :(return s)
+    exp = reduce((ex1,ex2) -> :(+($ex1, $ex2)), ex)
```
Comment (fredrikekre) [inline src/tensor_products.jl:55]: `reduce` here seems to be the only difference from `StaticArrays` compared to what we did before we built on that

File: src/tensor_products.jl [hunk 3]
```diff
@@ -251,24 +252,8 @@ end
     return Tensor{2, dim}(tomatrix(S1_t) * tomatrix(S2_t))
 end
 
-# Specialized methods for symmetric tensors
-@generated function Base.dot{dim}(S1::SymmetricTensor{2, dim}, v2::Vec{dim})
```
Comment (fredrikekre) [inline src/tensor_products.jl:255]: This was not even specialized. I still think its weird that its faster to convert to a `Tensor` and then use their `dot`...

File: src/tensor_products.jl [hunk 4]
```diff
@@ -345,9 +330,23 @@ julia> tdot(A)
  0.48726  0.540229  0.190334
 ```
 """
-@inline function tdot{dim}(S1::Tensor{2, dim})
-    return SymmetricTensor{2, dim}(transpdot(get_data(S1)))
+@generated function tdot{dim}(S1::Tensor{2, dim})
```
Comment (fredrikekre) [inline src/tensor_products.jl:333]: moved this inline, instead of having `transpdot` and `dot_matmatT`  in util files.

File: src/tuple_utils.jl [hunk 5]
```diff
@@ -1,29 +0,0 @@
-@generated function mat_get_index{N}(t::NTuple{N}, i::Int, j::Int)
```
Comment (fredrikekre) [inline src/tuple_utils.jl:1]: not used

File: src/utilities.jl [hunk 6]
```diff
@@ -22,14 +22,41 @@ macro code(e)
   isa(e, Expr) || error("can't generate code from non-expressions")
   return Expr(:escape, :(code = __append_code(code, $e)))
 end
-#########################################################################################
-
-function is_diagonal_index(dim::Int, i::Int)
```
Comment (fredrikekre) [inline src/utilities.jl:27]: not used

**Comment by KristofferC** [comment]
Net negative lines = Good PR. If benchmarks look good, good to go by me.

**Comment by fredrikekre** [comment]
only major benchmark diff is dcontract between two order 4sym. It wasn't inlined before.
Rest of the functions I have not really touched.

**Comment by fredrikekre** [comment]
Objections?

**Comment by KristofferC** [comment]
go

**Pull request merged**